File: qc-application/src/main/java/com/eager/questioncloud/application/business/creator/implement/CreatorStatisticsUpdater.java
package com.eager.questioncloud.application.business.creator.implement;

import com.eager.questioncloud.application.api.payment.question.event.QuestionPaymentEvent;
import com.eager.questioncloud.core.domain.creator.infrastructure.repository.CreatorStatisticsRepository;
import com.eager.questioncloud.core.domain.question.infrastructure.repository.QuestionRepository;
import com.eager.questioncloud.core.domain.question.model.Question;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class CreatorStatisticsUpdater {
    private final QuestionRepository questionRepository;
    private final CreatorStatisticsRepository creatorStatisticsRepository;

    @EventListener
    public void updateCreatorStatistics(QuestionPaymentEvent event) {
        List<Question> questions = questionRepository.getQuestionsByQuestionIds(event.getQuestionPayment().getOrder().getQuestionIds());
        Map<Long, Long> countQuestionByCreator = questions
            .stream()
            .collect(Collectors.groupingBy(Question::getCreatorId, Collectors.counting()));

        countQuestionByCreator
            .forEach((creatorId, count) -> creatorStatisticsRepository.addSalesCount(creatorId, count.intValue()));
    }
}

File: qc-application/src/main/java/com/eager/questioncloud/application/business/creator/implement/CreatorStatisticsUpdater.kt
package com.eager.questioncloud.application.business.creator.implement

import com.eager.questioncloud.application.business.payment.question.event.QuestionPaymentEvent
import com.eager.questioncloud.core.domain.creator.infrastructure.repository.CreatorStatisticsRepository
import com.eager.questioncloud.core.domain.question.infrastructure.repository.QuestionRepository
import com.eager.questioncloud.core.domain.question.model.Question
import io.awspring.cloud.sqs.annotation.SqsListener
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.stereotype.Component
import java.util.stream.Collectors

@Component
class CreatorStatisticsUpdater(
    private val questionRepository: QuestionRepository,
    private val creatorStatisticsRepository: CreatorStatisticsRepository,
) {
    @SqsListener("update-creator-sales-statistics.fifo")
    fun updateCreatorStatistics(@Payload event: QuestionPaymentEvent) {
        val questions = questionRepository.getQuestionsByQuestionIds(event.questionPayment.order.questionIds)
        val countQuestionByCreator = questions
            .stream()
            .collect(Collectors.groupingBy(Question::creatorId, Collectors.counting()))

        countQuestionByCreator
            .forEach { (creatorId: Long, count: Long) ->
                creatorStatisticsRepository.addSalesCount(
                    creatorId,
                    count.toInt()
                )
            }
    }
}
